src/benchmark/file_io_read_micro_benchmark.cpp

```diff
@@ -0,0 +1,205 @@
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <algorithm>
+#include <numeric>
+#include "micro_benchmark_basic_fixture.hpp"
+
+namespace hyrise {
+
+const int32_t MB = 1000000;
+
+class FileIOMicroReadBenchmarkFixture : public MicroBenchmarkBasicFixture {
+ public:
+  uint64_t control_sum = uint64_t{0};
+  std::vector<uint32_t> numbers;
+
+  void SetUp(::benchmark::State& state) override {
+    // TODO(everybody): Make setup/teardown global per file size to improve benchmark speed
+    ssize_t BUFFER_SIZE_MB = state.range(0);
+
+    // each int32_t contains four bytes
+    uint32_t vector_element_count = (BUFFER_SIZE_MB * MB) / 4;
+    numbers = std::vector<uint32_t>(vector_element_count);
+    for (size_t index = 0; index < vector_element_count; ++index) {
+      numbers[index] = std::rand() % UINT32_MAX;
+    }
+    control_sum = std::accumulate(numbers.begin(), numbers.end(), uint64_t{0});
+
+    int32_t fd;
+    if ((fd = creat("file.txt", O_WRONLY)) < 1) {
+      std::cout << "create error" << std::endl;
+    }
+    chmod("file.txt", S_IRWXU);  // enables owner to rwx file
+    if (write(fd, std::data(numbers), BUFFER_SIZE_MB * MB) != BUFFER_SIZE_MB * MB) {
+      std::cout << "write error" << std::endl;
+    }
+    close(fd);
+  }
+
+  void TearDown(::benchmark::State& /*state*/) override {
+    // TODO(everybody): Error handling
+    std::remove("file.txt");
+  }
+
+ protected:
+};
+
+BENCHMARK_DEFINE_F(FileIOMicroReadBenchmarkFixture, READ_NON_ATOMIC)(benchmark::State& state) {  // open file
+  int32_t fd;
+  if ((fd = open("file.txt", O_RDONLY)) < 0) {
+    std::cout << "open error " << errno << std::endl;
+  }
+  const int32_t NUMBER_OF_BYTES = state.range(0) * MB;
+
+  for (auto _ : state) {
+    state.PauseTiming();
+    micro_benchmark_clear_disk_cache();
+    std::vector<uint32_t> read_data;
+
+    read_data.resize(NUMBER_OF_BYTES / 4);
```

Comment:
I think we should use `reserve()` instead of `resize()`.
+ `sizeof()`

Reply:
Why an invalid state? From my understanding, reserve changes only the vector capacity (basically a realloc) but leaves the state (semantic) as it is.

Reply:
When you write data to `vector.data()`, the vector is not aware of any insertions. Thus, the following sum will not yield anything as end() has not been moved.
You would need to iteratore the allocation manually.

Reply:
Got it.
My point is that resize has to write new values if the vector gets larger.
This means, we have to write the vector twice.

Reply:
I can't follow. Why would the vector get larger?